File: src/ee/store/VirtualItemStorage.cpp
//
//  VirtualItemStorage.cpp
//  ee_x
//
//  Created by Zinge on 5/15/17.
//
//

#include <functional>

#include "ee/store/VirtualItemStorage.hpp"
#include "ee/store/VirtualItem.hpp"
#include "ee/store/KeyValueStorage.hpp"
#include "ee/store/StoreLogger.hpp"
#include "ee/store/StoreUtils.hpp"

#define ENSURE_NON_NEGATIVE_BALANCE 0

namespace ee {
namespace store {
namespace {
constexpr auto Tag = "EEX VirtualItemStorage";
} // namespace

int VirtualItemStorage::getBalance(VirtualItem* item) {
    auto itemId = item->getItemId();
    auto key = keyBalance(itemId);
    auto val = KeyValueStorage::getInstance().getValue(key);

    auto logDebug =
        std::bind(&StoreLogger::logDebug, std::ref(StoreLogger::getInstance()),
                  Tag, std::placeholders::_1);

    if (val.empty()) {
        logDebug(format("The balance for %s has not set yet, return zero",
                        itemId.c_str()));
        return 0;
    }

    auto balance = std::stoi(val);
#if ENSURE_NON_NEGATIVE_BALANCE
    if (balance < 0) {
        logDebug(format("The balance for %s is negative (%d), return zero",
                        itemId.c_str(), balance));
        return 0;
    }
#endif // ENSURE_NON_NEGATIVE_BALANCE

    logDebug(format("The balance for %s is %d", itemId.c_str(), balance));
    return balance;
}

int VirtualItemStorage::setBalance(VirtualItem* item, int balance,
                                   bool notify) {
    auto oldBalance = getBalance(item);

#if ENSURE_NON_NEGATIVE_BALANCE
    if (oldBalance < 0) {
        oldBalance = 0;
    }

    if (balance < 0) {
        balance = 0;
    }
#endif // ENSURE_NON_NEGATIVE_BALANCE

    if (oldBalance == balance) {
        return balance;
    }

    auto itemId = item->getItemId();
    auto key = keyBalance(itemId);
    auto val = std::to_string(balance);
    KeyValueStorage::getInstance().setValue(key, val);
    if (notify) {
        postBalanceChangeEvent(item, balance, balance - oldBalance);
    }
    return balance;
}

int VirtualItemStorage::addBalance(VirtualItem* item, int amount, bool notify) {
    auto balance = getBalance(item);

#if ENSURE_NON_NEGATIVE_BALANCE
    if (balance < 0) {
        balance = 0;
    }
#endif // ENSURE_NON_NEGATIVE_BALANCE

    if (amount == 0) {
        return balance;
    }

#if ENSURE_NON_NEGATIVE_BALANCE
    if (balance + amount < 0) {
        amount = -balance;
    }
#endif // ENSURE_NON_NEGATIVE_BALANCE

    auto itemId = item->getItemId();
    auto key = keyBalance(itemId);
    auto val = std::to_string(balance + amount);
    KeyValueStorage::getInstance().setValue(key, val);
    if (notify) {
        postBalanceChangeEvent(item, balance + amount, amount);
    }
    return balance + amount;
}
} // namespace store
} // namespace ee

File: src/android/soomla_store/src/main/java/com/soomla/store/data/VirtualItemStorage.java
/**
 * Copyright (C) 2012-2014 Soomla Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.soomla.store.data;

import com.soomla.SoomlaUtils;
import com.soomla.data.KeyValueStorage;

/**
 * This class is an abstract definition of a Virtual Item Storage.
 */
public abstract class VirtualItemStorage {

    /**
     * Retrieves the balance of the given virtual item.
     *
     * @param itemId the required virtual item's itemId
     * @return the balance of the required virtual item
     */
    public int getBalance(String itemId){
        SoomlaUtils.LogDebug(mTag, "fetching balance for virtual item with itemId: "
                + itemId);

        String key = keyBalance(itemId);
        String val = KeyValueStorage.getValue(key);

        int balance = 0;
        if (val != null) {
            balance = Integer.parseInt(val);
        }

        SoomlaUtils.LogDebug(mTag, "the balance for " + itemId + " is " + balance);

        return balance;
    }

    /**
     * Sets the balance of the given virtual item to be the given balance.
     *
     * @param itemId the required virtual item's itemId
     * @param balance the new balance to be set
     * @return the balance of the required virtual item
     */
    public int setBalance(String itemId, int balance) {
        return setBalance(itemId, balance, true);
    }

    /**
     * Sets the balance of the given virtual item to be the given balance, and if notify is true
     * posts the change in the balance to the event bus.
     *
     * @param itemId the required virtual item's itemId
     * @param balance the new balance to be set
     * @param notify if notify is true post balance change event
     * @return the balance of the required virtual item
     */
    public int setBalance(String itemId, int balance, boolean notify) {
        SoomlaUtils.LogDebug(mTag, "setting balance " + balance + " to " + itemId + ".");

        int oldBalance = getBalance(itemId);
        if (oldBalance == balance) {
            return balance;
        }

        String balanceStr = "" + balance;
        String key = keyBalance(itemId);

        KeyValueStorage.setValue(key, balanceStr);

        if (notify) {
            postBalanceChangeEvent(itemId, balance, 0);
        }

        return balance;
    }

    /**
     * Adds the given amount of items to the storage.
     *
     * @param itemId the required virtual item's itemId
     * @param amount the amount of items to add
     * @return new balance
     */
    public int add(String itemId, int amount){
        return add(itemId, amount, true);
    }

    /**
     * Adds the given amount of items to the storage, and if notify is true
     * posts the change in the balance to the event bus.
     *
     * @param itemId the required virtual item's itemId
     * @param amount the amount of items to add
     * @param notify if true posts balance change event
     * @return new balance
     */
    public int add(String itemId, int amount, boolean notify){
        SoomlaUtils.LogDebug(mTag, "adding " + amount + " " + itemId);

        int balance = getBalance(itemId);
        if (balance < 0) { /* in case the user "adds" a negative value */
            balance = 0;
            amount = 0;
        }
        String balanceStr = "" + (balance + amount);
        String key = keyBalance(itemId);
        KeyValueStorage.setValue(key, balanceStr);

        if (notify) {
            postBalanceChangeEvent(itemId, balance+amount, amount);
        }

        return balance + amount;
    }

    /**
     * Removes the given amount from the given virtual item's balance.
     *
     * @param itemId is the virtual item to remove the given amount from
     * @param amount is the amount to remove
     * @return new balance
     */
    public int remove(String itemId, int amount){
        return remove(itemId, amount, true);
    }

    /**
     * Removes the given amount from the given virtual item's balance, and if notify is true
     * posts the change in the balance to the event bus.
     *
     * @param itemId is the virtual item to remove the given amount from
     * @param amount is the amount to remove
     * @param notify if notify is true post balance change event
     * @return new balance
     */
    public int remove(String itemId, int amount, boolean notify){
        SoomlaUtils.LogDebug(mTag, "Removing " + amount + " " + itemId + ".");

        int balance = getBalance(itemId) - amount;
        if (balance < 0) {
            balance = 0;
            amount = 0;
        }
        String balanceStr = "" + balance;
        String key = keyBalance(itemId);
        KeyValueStorage.setValue(key, balanceStr);

        if (notify) {
            postBalanceChangeEvent(itemId, balance, -1*amount);
        }

        return balance;
    }

    /**
     * Retrieves the balance of the virtual item with the given itemId from the
     * <code>KeyValDatabase</code>.
     *
     * @param itemId id of the virtual item whose balance is to be retrieved
     * @return String containing name of storage base, itemId, and balance
     */
    protected abstract String keyBalance(String itemId);

    /**
     * Posts the given amount changed in the given balance of the given virtual item.
     *
     * @param itemId virtual item whose balance has changed
     * @param balance the balance that has changed
     * @param amountAdded the amount added to the item's balance
     */
    protected abstract void postBalanceChangeEvent(String itemId, int balance, int amountAdded);


    /** Private Members */

    protected String mTag = "SOOMLA VirtualItemStorage"; //used for Log messages
}
